### PR TITLE
[BLB] Implement Season of Gathering and Pawprints mechanic

### PR DESCRIPTION
**Pull request by jimga150**
#11853 
My current implementation requires pawprint cards to also set a min and max number of modes according to what logically could be chosen (0 and 5, for this and what will be the other four bloomburrow "Season of" cards).
In playtesting, the maxModes being set to 5 isnt required, since i added switches to check against the pawprint count instead of the mode count, but for some reason if i don't set the maxModes to 5 (default is 1), my tests all fail like this:

```
[ERROR] Failures: 
[ERROR]   PawPrintsTest.test_Choose113:36->CardTestPlayerAPIImpl.execute:296->CardTestPlayerAPIImpl.assertAllCommandsUsed:1599->CardTestPlayerAPIImpl.assertTargetsCount:1585 (Targets of PlayerA) Count are not equal (found [Memnite]). Some inner choose dialogs can be set up only in strict mode. expected:<0> but was:<1>
[ERROR]   PawPrintsTest.test_Choose122:212->CardTestPlayerAPIImpl.execute:296->CardTestPlayerAPIImpl.assertAllCommandsUsed:1598->CardTestPlayerAPIImpl.assertChoicesCount:1576 (Choices of PlayerA) Count are not equal (found [Artifact, Enchantment]). Some inner choose dialogs can be set up only in strict mode. expected:<0> but was:<2>
[ERROR]   PawPrintsTest.test_Choose123:70->CardTestPlayerAPIImpl.execute:296->CardTestPlayerAPIImpl.assertAllCommandsUsed:1598->CardTestPlayerAPIImpl.assertChoicesCount:1576 (Choices of PlayerA) Count are not equal (found [Enchantment]). Some inner choose dialogs can be set up only in strict mode. expected:<0> but was:<1>
[ERROR]   PawPrintsTest.test_Choose1x5:146->CardTestPlayerAPIImpl.execute:296->CardTestPlayerAPIImpl.assertAllCommandsUsed:1599->CardTestPlayerAPIImpl.assertTargetsCount:1585 (Targets of PlayerA) Count are not equal (found [Memnite, Memnite, Memnite, Memnite]). Some inner choose dialogs can be set up only in strict mode. expected:<0> but was:<4>
[ERROR]   PawPrintsTest.test_Choose2111:113->CardTestPlayerAPIImpl.execute:296->CardTestPlayerAPIImpl.assertAllCommandsUsed:1599->CardTestPlayerAPIImpl.assertTargetsCount:1585 (Targets of PlayerA) Count are not equal (found [Memnite, Memnite, Memnite]). Some inner choose dialogs can be set up only in strict mode. expected:<0> but was:<3>
[ERROR]   PawPrintsTest.test_Choose23:176->CardTestPlayerAPIImpl.execute:296->CardTestPlayerAPIImpl.assertAllCommandsUsed:1598->CardTestPlayerAPIImpl.assertChoicesCount:1576 (Choices of PlayerA) Count are not equal (found [Artifact]). Some inner choose dialogs can be set up only in strict mode. expected:<0> but was:<1>
[INFO] 
[ERROR] Tests run: 6, Failures: 6, Errors: 0, Skipped: 0
```

Which suggests that for some reason, the choice/target ingesting code is getting clogged. I'm not sure why, since i haven't had any luck trying to use breakpoints in maven run configurations.

### CODE REVIEW

File: Mage/src/main/java/mage/abilities/Mode.java
Comment (xenohedron) [inline Mage/src/main/java/mage/abilities/Mode.java:22]: please add comment on usage of "magic number"

File: Mage/src/main/java/mage/abilities/Mode.java
Comment (xenohedron) [inline Mage/src/main/java/mage/abilities/Mode.java:125]: what's the point of having separate "set" and "with" methods? I would just have the "with" method, the return can be ignored if not needed

File: Mage/src/main/java/mage/abilities/Modes.java
Comment (xenohedron) [inline Mage/src/main/java/mage/abilities/Modes.java:58]: comment here as well

File: Mage/src/main/java/mage/abilities/Modes.java
Comment (xenohedron) [inline Mage/src/main/java/mage/abilities/Modes.java:462]: don't throw RuntimeException, use something like IllegalStateException

File: Mage/src/main/java/mage/abilities/Modes.java
Comment (xenohedron) [inline Mage/src/main/java/mage/abilities/Modes.java:471]: what's up with this change?

File: Mage/src/main/java/mage/abilities/dynamicvalue/common/ControlledCreaturesGreatestPower.java
Comment (xenohedron) [inline Mage/src/main/java/mage/abilities/dynamicvalue/common/ControlledCreaturesGreatestPower.java:13]: GreatestPowerAmongControlledCreaturesValue already exists

(and is a good example of stream, mapToInt, sum as a cleaner style)

**Review by xenohedron — COMMENTED** [review]
Overall this looks good to me! I'll let another dev take a look before merging. Thanks for the test cases.

File: Mage.Sets/src/mage/cards/s/SeasonOfGathering.java
Comment (JayDi85) [inline Mage.Sets/src/mage/cards/s/SeasonOfGathering.java:53]: Why paw logic done by mode’s min/max style instead mode cost? Is it a good design here for it? @theelk801

Comment (xenohedron) [reply Mage.Sets/src/mage/cards/s/SeasonOfGathering.java:53]: We don't have the CR update yet, but {P} is not a cost.

From [mechanics article](https://magic.wizards.com/en/news/feature/bloomburrow-mechanics)
> As you cast one of these modal spells, you can choose up to five pawprints worth of modes. Pawprints are a new symbol used exclusively for these cards. They aren't a resource like mana or energy. You can't save them or carry them over from one spell to the next.

So it seems to me that this is appropriate design.

**Review by JayDi85 — COMMENTED** [review]
Require some code style and check logic changes. See related comments.

File: Mage.Sets/src/mage/cards/s/SeasonOfGathering.java
Comment (JayDi85) [reply Mage.Sets/src/mage/cards/s/SeasonOfGathering.java:53]: Then it’s fine

File: Mage/src/main/java/mage/abilities/Modes.java
Comment (jimga150) [reply Mage/src/main/java/mage/abilities/Modes.java:471]: I'm using the Mode obtained from `get(modeId)` multiple times so i bumped it to a local variable

File: Mage/src/main/java/mage/abilities/Modes.java
Comment (JayDi85) [reply Mage/src/main/java/mage/abilities/Modes.java:462]: For info:
- verify check -- can be called from card constructor (by card, ability, effect or other data setup). There are some verify tests that call constructor for all cards and tokens. So such errors will be catches by tests as fast as possible.
- runtime check -- can be called from real game only.

File: Mage.Server.Plugins/Mage.Player.AI/src/main/java/mage/player/ai/ComputerPlayer.java
Comment (jimga150) [reply Mage.Server.Plugins/Mage.Player.AI/src/main/java/mage/player/ai/ComputerPlayer.java:2220]: `chooseMode` is where the max modes check is, so it seemed symmetrical to place the max pawprints check here too. should the max modes check move to `getAvailableModes` too?

File: Mage.Server.Plugins/Mage.Player.Human/src/mage/player/human/HumanPlayer.java
Comment (jimga150) [reply Mage.Server.Plugins/Mage.Player.Human/src/mage/player/human/HumanPlayer.java:2575]: Linking for my reference https://github.com/magefree/mage/pull/12617#discussion_r1696331613

File: Mage.Tests/src/test/java/org/mage/test/player/TestPlayer.java
Comment (jimga150) [reply Mage.Tests/src/test/java/org/mage/test/player/TestPlayer.java:2085]: Linking for my reference https://github.com/magefree/mage/pull/12617#discussion_r1696331613

File: Mage.Server.Plugins/Mage.Player.AI/src/main/java/mage/player/ai/ComputerPlayer.java
Comment (JayDi85) [reply Mage.Server.Plugins/Mage.Player.AI/src/main/java/mage/player/ai/ComputerPlayer.java:2220]: Try to add paw check to getMaxMode. getAvailableModes must use paw check too, not getMaxMode.

File: Mage/src/main/java/mage/abilities/Modes.java
Comment (jimga150) [reply Mage/src/main/java/mage/abilities/Modes.java]: To clarify, are you askign for both here? I moved the runtime check pointed to in this thread to a check in addMode() to catch it earlier.

**Comment by jimga150** [comment]
Is there a way to retrigger travis builds? this one looks like it was a transient error

**Review by xenohedron — COMMENTED** [review]
minor followup correction needed, otherwise I think this is ready to go

File: Mage.Server.Plugins/Mage.Player.Human/src/mage/player/human/HumanPlayer.java
Comment (jimga150) [reply Mage.Server.Plugins/Mage.Player.Human/src/mage/player/human/HumanPlayer.java:2612]: yeah, missed this and another check when converting the pattern to 0 -> non-pawprints

**Review by xenohedron — APPROVED** [review]

**Pull request merged**
